### PR TITLE
Add file size to FileOptions

### PR DESCRIPTION
**Pull request by acvictor**
Allow specifying the file length via `FileOptions`.
When opening a file for reading, this helps avoid a call to remote storage to fetch file metadata. 
This is one of the changes needed to support https://github.com/facebookincubator/velox/issues/9314

### CODE REVIEW

**Comment by acvictor** [comment]
@majetideepak can you let me know what you think of this?

**Comment by acvictor** [comment]
> @majetideepak can you let me know what you think of this?

I think the intent behind adding infoColumns is to add metadata that is being queried but I want to always pass length and modification time to the reader so that the call to remote storage to fetch properties is not made. I think it might actually a better idea to add a new field. What do you think? I see that in Gluten also this metadata is passed down to the split only depending on the output of the scan spec https://github.com/apache/incubator-gluten/blob/0bfee3a981797d78e8598ede3ea09854e0718ff8/shims/spark34/src/main/scala/org/apache/spark/sql/execution/FileSourceScanExecShim.scala#L53

**Review by majetideepak — COMMENTED** [review]
@acvictor One high-level comment.
I will complete the review by EOD. Thanks!

File: velox/connectors/hive/FileHandle.h
Comment (acvictor) [reply velox/connectors/hive/FileHandle.h:74]: File size/modification time need to be passed to openFileForRead and so need to be avialable here https://github.com/facebookincubator/velox/blob/d36507681b23a3beae1698fa4b1979dcb5199591/velox/connectors/hive/FileHandle.cpp#36.

Comment (majetideepak) [reply velox/connectors/hive/FileHandle.h:74]: I mean't we can add the `FileHandleFactory` change after this PR lands.
I am running behind on this due to travel/prep to VeloxCon. Sorry for the delay. I want to look at a couple of other APIs first.

Comment (acvictor) [reply velox/connectors/hive/FileHandle.h]: Thanks @majetideepak. I've removed changes related to `FileHandleFactory`.

File: velox/connectors/hive/HiveConnectorUtil.h
Comment (majetideepak) [inline velox/connectors/hive/HiveConnectorUtil.h:38]: unused?

File: velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp:77]: Let's add ` static constexpr folly::StringPiece kFileSize{"file-size"};` to the `FileSize` class.
We should throw an error if `options.values.at("fileSize").empty()` is true as it is a user error.

**Review by majetideepak — COMMENTED** [review]
@acvictor thanks for this PR! A couple of comments.

**Comment by majetideepak** [comment]
Can you also add a description to this PR?

File: velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp
Comment (acvictor) [reply velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp]: I see what you mean, updated. Thanks!

**Review by majetideepak — COMMENTED** [review]
@acvictor The changes look good. I feel adding an API to FileOptions say
`std::optional<size_t> getFileSize()` is probably cleaner.
 Can you add a test as well? If we add the API, then it should be sufficient to add one test to the `S3FileSystemTest.cpp`
`std::stoull` can throw. Let's add a test for that as well.

File: velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp
Comment (majetideepak) [reply velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp]: Yes, `FileOptions` struct. You got this right!

**Comment by majetideepak** [comment]
@acvictor I now think we could add a field `std::options<size_t> fileSize` to `FIleOptions` and not have to deal with string conversion. We can rename `values` to `extraValues` and update the code comments.
The other question I have is that the `FileHandleGenerator::operator()` does not specify `FileOptions` when invoking `openFileForRead`. Do we have to modify `FileHandleGenerator` APIs as well?

**Comment by acvictor** [comment]
> @acvictor I now think we could add a field `std::options<size_t> fileSize` to `FileOptions` and not deal with string conversion. We can rename `values` to `extraValues` and update the code comments. The other question I have is that the `FileHandleGenerator::operator()` does not specify `FileOptions` when invoking `openFileForRead`. Do we have to modify `FileHandleGenerator` and up the call stack APIs as well?

Thanks @majetideepak. I've updated `FileOptions` to include `std::optional<int64_t> fileSize` and a function `int64_t getFileSize() const`.

Also, yes. `FileHandleGenerator` needs to be updated along with the up the call stack APIs (I had these in an earlier commit where I'd made it a tuple).

File: velox/common/file/FileSystems.h
Comment (majetideepak) [inline velox/common/file/FileSystems.h:48]: `std::optional<int64_t> getFileSize()` Let's return the optional value directly here.

File: velox/connectors/hive/storage_adapters/abfs/AbfsFileSystem.cpp
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/abfs/AbfsFileSystem.cpp:64]: If the value `options.getFileSize()` is present but is ` < 0`, then we need to throw an error here since the expectation in this method is that `length_ >= 0`.

**Review by majetideepak — COMMENTED** [review]
@acvictor This looks much cleaner. We should still throw a user error if the input specified is negative.

**Comment by majetideepak** [comment]
> Also, yes. `FileHandleGenerator` needs to be updated along with the up the call stack APIs (I had these in an earlier commit where I'd made it a tuple).

I see that now. Let's add this in a separate PR.

File: velox/connectors/hive/storage_adapters/abfs/AbfsFileSystem.cpp
Comment (majetideepak) [reply velox/connectors/hive/storage_adapters/abfs/AbfsFileSystem.cpp:64]: Please add a test for this.

**Review by majetideepak — APPROVED** [review]
Thanks, @acvictor

**Comment by acvictor** [comment]
> Thanks, @acvictor

Thank you @majetideepak for the review! @mbasmanova can you please merge this if it looks okay?

File: velox/common/file/FileSystems.cpp
Comment (mbasmanova) [inline velox/common/file/FileSystems.cpp:89]: You still need to put 'options' in comments to avoid linter warning as unused parameter.

File: velox/common/file/FileSystems.h
Comment (mbasmanova) [inline velox/common/file/FileSystems.h:48]: What's the point of adding this getter if 'fileSize' is public?

File: velox/connectors/hive/storage_adapters/abfs/AbfsFileSystem.cpp
Comment (mbasmanova) [inline velox/connectors/hive/storage_adapters/abfs/AbfsFileSystem.cpp:67]: nit: Length -> File size

File: velox/connectors/hive/storage_adapters/abfs/AbfsFileSystem.cpp
Comment (mbasmanova) [inline velox/connectors/hive/storage_adapters/abfs/AbfsFileSystem.cpp:66]: nit: if (auto fileSize = options.getFileSize()) {...}

File: velox/common/file/FileSystems.h
Comment (mbasmanova) [inline velox/common/file/FileSystems.h:46]: This doesn't quite feel like an option. Would you add a comment to explain that if this is specified, then this value can be trusted?

File: velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp
Comment (mbasmanova) [inline velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp:85]: same comments as above

File: velox/connectors/hive/storage_adapters/hdfs/HdfsFileSystem.cpp
Comment (mbasmanova) [inline velox/connectors/hive/storage_adapters/hdfs/HdfsFileSystem.cpp:72]: same as above

File: velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp
Comment (mbasmanova) [inline velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp:78]: ditto

File: velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp
Comment (mbasmanova) [inline velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp:178]: How is this test verifying the changes? Would it fail without the changes or if we remove code that processes options.fileSize?

File: velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp
Comment (mbasmanova) [inline velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp:186]: nit: no need to create a shared pointer

```
filesystems::abfs::AbfsFileSystem abfs{hiveConfig};

abfs.openXxx
```

**Review by mbasmanova — COMMENTED** [review]
@acvictor Some comments.

File: velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp
Comment (acvictor) [reply velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp]: The test will not fail if we remove the code that processes `options.sizeSize`. However, since fileSize is not 0, the `ASSERT_EQ(readFile->size(), 15 + kOneMB)` in `readData` should test that the expected file size is being set.

File: velox/common/file/FileSystems.cpp
Comment (acvictor) [reply velox/common/file/FileSystems.cpp]: Updated, thanks!

File: velox/common/file/FileSystems.h
Comment (acvictor) [reply velox/common/file/FileSystems.h]: Removed, this was earlier returning the optional's value.

File: velox/connectors/hive/storage_adapters/abfs/AbfsFileSystem.cpp
Comment (acvictor) [reply velox/connectors/hive/storage_adapters/abfs/AbfsFileSystem.cpp]: Updated everywhere.

File: velox/common/file/FileSystems.h
Comment (acvictor) [reply velox/common/file/FileSystems.h]: Added a comment.

File: velox/connectors/hive/storage_adapters/hdfs/HdfsFileSystem.cpp
Comment (acvictor) [reply velox/connectors/hive/storage_adapters/hdfs/HdfsFileSystem.cpp]: Updated.

File: velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp
Comment (acvictor) [reply velox/connectors/hive/storage_adapters/abfs/tests/AbfsFileSystemTest.cpp]: Updated everywhere.

File: velox/common/file/FileSystems.h
Comment (mbasmanova) [inline velox/common/file/FileSystems.h:46]: - no need to explicitly initialize to std::nullopt
- comment goes on a line above the property; use /// for public APIs

```
/// If specified then can be trusted to be the file size.
std::optional<int64_t> fileSize;
```

**Review by mbasmanova — APPROVED** [review]
@acvictor Looks good % a nit.

**Comment by acvictor** [comment]
> @acvictor Looks good % a nit.

Thank you @mbasmanova for the review!

**Review by Yuhta — APPROVED** [review]

File: velox/common/file/FileSystems.h
Comment (assignUser) [reply velox/common/file/FileSystems.h]: It seems the compiler disagrees: https://github.com/facebookincubator/velox/actions/runs/8636921920/job/23677942499?pr=9313#step:7:3240

Comment (assignUser) [reply velox/common/file/FileSystems.h]: nvm I didn't see the follow up commit

Comment (acvictor) [reply velox/common/file/FileSystems.h]: You're right, it was failing. Raised a fix here https://github.com/facebookincubator/velox/pull/9441.

**Pull request closed without merging**
